### ISSUE TITLE
Use more consistent taxonomy and APIs for representing arity.

### ISSUE THREAD

**Issue by olson-sean-k** [olson-sean-k/plexus#42]
Plexus uses the term _arity_ to refer to the edge count of polygons. There are two orthogonal concerns that are sometimes conflated or otherwise unclear in APIs:

- The different natures of the arity of primitive types like `NGon` and mesh types like `MeshGraph`. At runtime, primitive types have a singular arity while meshes may have a compound arity formed from their parts (this could be a singular value or a range of values).
- The arities that a type statically supports vs. the arity that a value has at runtime. This can be modeled with an associated constant and an arity function over `&self`, respectively.

For example, the `ConstantArity` trait is implemented by primitive types that have a singular arity and so arity is expressed as a single `usize`. In this case, static and dynamic arity are implied by this trait: they are the same singular value. But what about dynamic types and mesh types? At the moment, this is not described by a trait and the naming would likely clash with the `ConstantArity` trait and `Arity` type.

It would probably be good to codify this and use a consistent taxonomy in APIs. It may be helpful for types to be able to specify an associated type for representing arity that in turn implements some common traits. Perhaps something like the following:

```rust
pub trait StaticArity {
    type Static;

    const ARITY: Self::Static;
}

pub trait DynamicArity: StaticArity {
    type Dynamic; // Distinct from `Static`. Consider the `Polygon` type.

    fn arity(&self) -> Self::Dynamic;
}

// Implemented by primitive types with fixed arity, like `NGon([T; 3])`. Bikeshed...
pub trait ConstantArity: StaticArity<Static = usize> {}

// Prevent alternative implementations of `DynamicArity` for types that implement
// `ConstantArity`.
impl<T> DynamicArity for T
where
    T: ConstantArity,
{
    type Dynamic = <T as StaticArity>::Static;

    fn arity(&self) -> Self::Dynamic {
        Self::ARITY
    }
}
```

Of course, documenting this would also be necessary.

**Comment by olson-sean-k** [comment]
I've experimented with this in ef2143a. The arrangement of types and traits is a bit different:

```rust
pub trait Arity: Copy {
    fn into_interval(self) -> (usize, Option<usize>);
}

pub trait StaticArity {
    type Static;

    const ARITY: Self::Static;
}

pub trait DynamicArity: StaticArity {
    type Dynamic;

    fn arity(&self) -> Self::Dynamic;
}

pub trait Homomorphic: StaticArity<Static = usize> {}
```

Here, `Homomorphic` replaces `ConstantArity` and is implemented by types that support a single arity (of type `usize`). This includes `NGon` and certain `MeshBuffer` types, as opposed to polymorphic types like `Polygon` and `MeshGraph`.

The following table breaks down the types implementing `Arity`:

|              | Static                      | Dynamic          |
|--------------|-----------------------------|------------------|
| **Singular** | `(usize, usize)` or `usize` | `usize`          |
| **Compound** | `(usize, Option<usize>)`    | `(usize, usize)` |

This table distinguishes between _static_ vs. _dynamic_ and _singular_ vs. _compound_. In the singular-static case, it is also important to consider _homomorphic_ vs. _polymorphic_. The `Homomorphic` trait operates at a type level: arity for these types is **always** represented by `usize` and never differs from `StaticArity::ARITY`. Moreover, `Homomorphic` types implementing `DynamicArity` must ensure that `self.arity()` and `Self::ARITY` are always equal.

The `DynamicArity` trait is included in the `prelude` module and replaces inherent `arity` functions.

**Comment by olson-sean-k** [comment]
These changes landed on `master` in 231a45f. More documentation will be needed, but I'm considering this closed by that change.
